### DISCUSSION TITLE
Templating: Change dashboard time range by selecting a template variable value

### DISCUSSION THREAD

**Discussion by rovegrod** [grafana/grafana#38138] [Dashboards, Panels & Visualizations]
Currently, we can make 'hand-crafted' queries using Influx as data-source and add inside templates-variables as time range boundaries, as follow: 
`SELECT mean("value") FROM "metric" WHERE time >=${from}ms and time <=${to}ms  GROUP BY time($__interval) `
Where $from and $to are epoch values and whose value can change depending on, for example, another template-variable.

If I've selected a quick range on the global/absolute time-range panel in Grafana like 'Yesterday' and I change the value of my variables we can see in the panels a **Zoom to data** option:
![Screenshot 2020-08-13 at 12 44 38](https://user-images.githubusercontent.com/50721154/90126057-507b9a80-dd63-11ea-98ba-44d4a289cb01.png)
The queries are working ok using my custom variables var-from and var-to. The problem I'm facing is that the visible "time-frame" shown by the Graph panels is determined by the global time range selected in grafana (i.e $__from and $__to), and since I am loading data from a different time range, the points are not "visible" in the graph.
I have verified that when the time range of the data I get (determined by var-from and var-to) is prior to the time range selected in grafana (determined by $__from and $__to), the panel shows a "Zoom to data" button that lets the user move to the visible "time-frame" in which there are points. However, when the time range of my variables is later in time than the time range selected in grafana, this button does not appear.

**What would you like to be added**:

I'd like grafana to show the "Zoom to data" button and let me "position" to the visible "time-frame" in any case in which there is data outside the visible "time-frame", no matter whether the data is positioned before or after the visible/currently selected range.

**Comment by torkelo** [comment]
Sorry I do not understand what your asking for, can you describe your scenario in more detail, what problem & workflow are you trying to solve and what is not working or what would like Grafana to support / do.

**Comment by rovegrod** [comment]
Hi @torkelo,

I'm sorry I did no make myself clear before.

I need to update the global time range that is applied to a dashboard according to the value of a couple of template-variables that I have (`var-from` and `var-to`).
I fill these custom variables with a time range in epoch format from a query to an InfluxDB database and later use that stored values to do other queries.

The queries are working ok using my custom variables `var-from` and `var-to`. The problem I'm facing is that the visible "time-frame" shown by the Graph panels is determined by the global time range selected in grafana (i.e `$__from` and `$__to`), and since I am loading data from a different time range, the points are not "visible" in the graph.

I have verified that when the time range of the data I get (determined by `var-from` and `var-to`) is prior to the time range selected in grafana (determined by `$__from` and `$__to`), the panel shows a "Zoom to data" button that lets the user move to the visible "time-frame" in which there are points. However, when the time range of my variables is later in time than the time range selected in grafana, this button does not appear.

I'd like grafana to show the "Zoom to data" button and let me "position" to the visible "time-frame" in any case in which there is data outside the visible "time-frame", no matter whether the data is positioned before or after the visible/currently selected range.

I hope I have explained myself a little better this time and thank you very much for responding so quickly to the request.
You're making a great product there!

**Comment by jdmaloney** [comment]
@torkelo Adding a "+1" to this as well, trying to find a work around to this myself (just with times from a mysql query being stored as template variables).  

Use case is HPC job metrics.  Building a dashboard where user plugs in their Job ID that ran on the system; that job id is used to query the mysql database to get the hosts the job ran on and the start and end time.  Using those results in my queries to build the dashboard panels that show job cpu/mem/gpu/disk/network profiles, etc.  Would love be able to set the dashboard time range to the results of the mysql queries that generates the start and end times. So the graphs zoom to the job's time window.

**Comment by torkelo** [comment]
I think the way to solve this would be to flag a template variable as representing a time range and when you change the variable it would change the dashboard time range. So it would be a template variable where the options would have a from & to (and maybe a text representation). 

Would this work for you?

**Reply by JoeH20** [reply]
How would we go about this?

**Reply by beasteers** [reply]
A possibly simpler solution would be to allow variable expansion in the time range field. The added benefit to this is that it would not be constrained to setting the global time range (you could use it in a repeater and on individual panels for example)

Additionally you'd be able to arbitrarily do: `from: $some_timestamp-10m` + `to: $some_timestamp+10m` to show the time around some event

**Reply by vaibssharma007** [reply]
hi @torkelo any update on this? is this feature introduced yet?

**Reply by Trolldemorted** [reply]
@torkelo after all this time, I think just **any** solution that enables us to do something like this would be good enough.

**Reply by up201306506** [reply]
Unfortunate that this isn't yet possible. Would make selecting the time frame for a load test easier. I kind of expected that we could just use variables on the custom time range fields at least.

Seems like the only workaround is creating a url with the `from=` and `to=` parameters set, but even that brings up a weird question - how come we can use absolute timestamps on the url, but pasting one in the timeframe window just yields a complaint from the UI? We can't even copy paste a query's result value ourselves without having to muck around with an address bar.

I guess the least-hindrance solution is to just have a link in the dashboard that doubles back into itself. That sounds very hacky as it demands page reload. But at least it works.  `/d/${__dashboard.uid}/${__dashboard}?from=${timeStart}&to=${timeEnd}` for those interested.

**Comment by jdmaloney** [comment]
@torkelo Yes I think that would be just fine.  Leveraging the template variable functionality makes sense.

**Comment by vellala2000** [comment]
+1 for this feature. I am also trying to update dashboard time based on variable pulled from influx query.

Seems like pretty reasonable expectation. 

Any reason why grafana not letting update/set global variables dynamically in template based on user interaction. It seems to be taking $variable literally when set in {time: to:$variable} unlike in queries

**Comment by vellala2000** [comment]
@rovegrod How are you able to read timestamp from influxDB query to grafana variable. I went through a bit of struggle to read time stamp instead of value as influxql query not allow to read timestamp alone without value. I was able to do it with flux query, just curious

**Comment by dmiego** [comment]
+1 for this feature

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by Sulomus** [comment]
Did anyone find a way how to do this?

**Comment by arodriguezg98** [comment]
+1

**Comment by jlopez0** [comment]
+1

**Comment by vuthongpadi** [comment]
+1

**Comment by raudie** [comment]
+1

**Comment by natellium** [comment]
Hey there, we're migrating back to Gh issues so I've created one to keep track of this https://github.com/grafana/grafana/issues/78527

**Comment by dorinpirvu** [comment]
+1 for this feature. still relevant in 2024.
@rovegrod, i'm also interested in how do you build your time variables - my (variable-)queries return values in the form 1.something e+12, which is not correctly interpreted by the in-panel queries... (e.g. specifiying "WHERE time >= ${StartTime:ms} and time <= ${EndTime:ms}" would return "InfluxDB returned error: error parsing query: found e, expected ; at line 5, char 29" - that's the exact position of the e from the variable) - thanks in advance if you ever see this question and respond :)
